### ISSUE TITLE
Modal form created $Ajax issue

### ISSUE THREAD

**Issue by natGalynker** [ga-wdi-boston/game-project#219]
When I created the modal for sign-up/sign-in something changed with $AJAX. I think what needs to happen is that the corresponding JS for the modal probably needs to move into the js file with the event handlers. But I am unsure about the right order to do things in that file now. If the $(document).ready(function(){ should come after module exports or before. If it should be inside the handlers function. Or right after. Part of the modal uses on clicks. Which is part of the confusing part for me currently.

**Comment by laurenfazah** [comment]
One thing at a time here. What changed with AJAX when you created the modal? As in what was happening before and what is happening now?

**Comment by gaand** [comment]
I don't think any of the rearrangement you're contemplating is necessary.

**Comment by natGalynker** [comment]
Before the modal I in the console I was receiving GET, POST and PATCH request from the server. Now Even before I click a button there is a get on the screen linking to here : http://localhost:8080/sockjs-node/info?t=1469582776261" when I enter sign-up information into the input fields I created with the modal I get undefined. It also doesnt return the json object with the information I entered in the input fields but prints out the name of the modal I am operating on

**Comment by laurenfazah** [comment]
Can I see your event handlers being passed to document ready?

**Comment by laurenfazah** [comment]
Also, could you post the sign-up form, and all related JS leading to the AJAX request?

**Comment by natGalynker** [comment]
I tried changing some of the naming of Ids and classes in the html so that they would have the same names as the relating html but that just broke the modal. I think I was on the right track but here is the relevant code

```
<nav class="navbar navbar-default">
    <div class="container-fluid">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#signUpModal">Sign UP</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#signInModal">Sign In</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#changePwModal">Change Password</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#changePlayerModal">Change Player</button>
        <a class="navbar-brand" href="#">Tic Tac Toe</a>
      </div>
    </div>
  </nav>
  <div class="modal fade" id="signUpModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
          <h4 class="modal-title" id="myModalLabel">Sign Up</h4>
        </div>
        <div class="modal-body">
          <div class= "input-group">
            <form class="form-group" id="signUpModal">
              <input type="text" class="form-control sign-in"
              name="credentials[player_name" placeholder="player name">
              <input type="text" class="form-control email"
              name="credentials[email]" placeholder="e-mail">
              <input type="password" class="form-control password"
              name="credentials[password]" placeholder="password">
              <input type="password" class="form-control confirm-pw"
              name="credentials[password_confirmation]"
              placeholder="confirm password">
            </form>
          </div>
        </div>
        <div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
          <button type="button" class="this-test-modal btn btn-primary">Sign Up!</button>
        </div>
      </div>
    </div>
```

```
'use strict';
const getFormFields = require(`../../../lib/get-form-fields`);


const api = require('./api');
const ui = require('./ui');

const onSignUp = function (event) {
  let data = getFormFields(this);
  event.preventDefault();
  api.signUp(data)
  .done(ui.success)
  .fail(ui.failure);
};
const addHandlers = () => {
  $('#sign-up').on('submit', onSignUp);
  $('#sign-in').on('submit', onSignIn); //grab element from the dom with element
  $('#change-password').on('submit', onChangePassword);
  $('#sign-out').on('submit', onSignOut);    //id on sign-up. Then does something
  $('#change-player').on('submit', onChangePlayer);  //with the id it grabbed.
  $('#new-game').on('click', onNewGame);

};
$(() => {
  $('.space').on('click', function(){
    console.log('it clicked!');
  });


});
module.exports = {
  addHandlers,
//only use module.exports on variables which are needed to be accessed outside
//of that file
};

const signUp =  function(data) {
  console.log(data);
return $.ajax({ //same as return $.ajax({
    url: app.api + '/sign-up',
    method: 'POST',
    data,
});
};

#(() => {
  authEvents.addHandlers(); //whatever is in here is what gets run
  $('#signUpModal').on('click', function () {
    $('#signUpModal').modal('show');
});
});
```

**Comment by laurenfazah** [comment]
A few things I see:
1. id `signUpModal` is being used twice. That's not allowed in HTML. 
2. You have an event handler watching for a form with `#sign-up`'s submission, but your form itself is not carrying that ID. 

Please comb through your event handlers and markup to make sure everything is aligned with what you're trying to connect (in terms of DOM elements).

**Comment by natGalynker** [comment]
I tried doing that just now, when I click the button to submit, nothing happens, it doesn't refresh it just stays there as is. I have tried tweaking. I made sure that it was a submit instead of a button, I tried connecting everything I had working from the rich HTML form over, but I am unsure what I am missing

**Comment by laurenfazah** [comment]
Could you post the HTML you used to create the submit button?

**Comment by natGalynker** [comment]
<div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
          <button type="submit" name= "submit" class="signUp btn btn-primary">Sign Up!</button>
        </div>
      </div>

**Comment by natGalynker** [comment]
oops 
```   <div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
          <button type="submit" name= "submit" class="signUp btn btn-primary">Sign Up!</button>
        </div>
      </div>
```

**Comment by natGalynker** [comment]
```
<div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
          <button type="submit" name= "submit" class="signUp btn btn-primary">Sign Up!</button>
        </div>
      </div>
```

**Comment by laurenfazah** [comment]
What type of HTML element have we been using to submit forms (using the `type="submit"` attribute)?

**Comment by natGalynker** [comment]
input That was how I was setting up my sign-in and etc forms

**Comment by laurenfazah** [comment]
Perfect - you should want to use the `<input>` element for sign-up, just as you did with the other forms. It is the element (when `type="submit"` is present) that will fire a form submit event.

**Comment by natGalynker** [comment]
It didn't work. It is still at a standstill.

**Comment by laurenfazah** [comment]
Could you post the new submit `<input>` element?

**Comment by natGalynker** [comment]
```
    <div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
          <input type="submit" name="submit" value="Sign up!">
```

**Comment by laurenfazah** [comment]
That looks good. Could you please post all relevant HTML for the form as well as corresponding JS for sign up (same as before)?

I'd like to see what the markup and event handlers look like once their IDs were sorted out.

**Comment by natGalynker** [comment]
```
<nav class="navbar navbar-default">
    <div class="container-fluid">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#signUpModal">Sign UP</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#signInModal">Sign In</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#changePwModal">Change Password</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#changePlayerModal">Change Player</button>
        <a class="navbar-brand" href="#">Tic Tac Toe</a>
      </div>
    </div>
  </nav>
  <div class="modal fade" id="signUpModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
          <h4 class="modal-title" id="myModalLabel">Sign Up</h4>
        </div>
        <div class="modal-body">
          <div class= "input-group">
            <form class="form-group" id="sign-up">
              <input type="text" class="form-control sign-in"
              name="credentials[player_name" placeholder="player name">
              <input type="text" class="form-control email"
              name="credentials[email]" placeholder="e-mail">
              <input type="password" class="form-control password"
              name="credentials[password]" placeholder="password">
              <input type="password" class="form-control confirm-pw"
              name="credentials[password_confirmation]"
              placeholder="confirm password">
            </form>
          </div>
        </div>
        <div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
          <input type="submit" name="submit" value="Sign up!">
        </div>

      </div>
    </div>
```

```
$(() => {
  authEvents.addHandlers(); //whatever is in here is what gets run
  $('#sign-up').on('click', function () {
    $('#signUpModal').modal('show');
});
});
```

```
const onSignUp = function (event) {
  let data = getFormFields(this);
  event.preventDefault();
  api.signUp(data)
  .done(ui.success)
  .fail(ui.failure);
};
const addHandlers = () => {
  $('#sign-up').on('submit', onSignUp);
  $('#sign-in').on('submit', onSignIn); //grab element from the dom with element
  $('#change-password').on('submit', onChangePassword);
  $('#sign-out').on('submit', onSignOut);    //id on sign-up. Then does something
  $('#change-player').on('submit', onChangePlayer);  //with the id it grabbed.
  $('#new-game').on('click', onNewGame);

};

const signUp =  function(data) {
  console.log(data);
return $.ajax({ //same as return $.ajax({
    url: app.api + '/sign-up',
    method: 'POST',
    data,
});
};
```

**Comment by laurenfazah** [comment]
Does having this:

``` js
  $('#sign-up').on('click', function () {
    $('#signUpModal').modal('show');
});
```

and this:

``` js
$('#sign-up').on('submit', onSignUp);
```

look right to you?

I would guess that you meant to capture the Sign Up button on click to show the #signUpModal

**Comment by natGalynker** [comment]
Yes, that was what my intent was, I that that those two needed to have the same naming conventions to work.

I just changed the id of the modal to SignUp. so the js looks like this now:

```
  $('#sign-up').on('submit', onSignUp);

authEvents.addHandlers(); //whatever is in here is what gets run
  $('#signUpModal').on('click', function () {
    $('#signUp').modal('show');
});
});
```

now its not launching at all again.

**Comment by laurenfazah** [comment]
If it was working before, I would suggest changing it back to that state, and changing the ID of your sign-up form to something unique. 

Let me know what happens there and I can assess if this needs an in person diagnosis.

**Comment by natGalynker** [comment]
It launched, I changed the form's ID and the handler to 

```
$('#player-sign-up').on('submit', onSignUp);
```

it still is not sending the information. Just stays in the form state without sending it or refreshing

**Comment by laurenfazah** [comment]
Want to come to the dugout to diagnose?

**Comment by laurenfazah** [comment]
@natGalynker could you summarize our solution and close please? Thanks!

**Comment by natGalynker** [comment]
The input field was outside of the form tag. I had to reorganize the html to get it to communicate properly

**Comment by gaand** [comment]
@natGalynker This appears to be resolved.  If so, would you please close?  Thanks.

**Comment by natGalynker** [comment]
On my navbar I have 3 buttons: sign-up, sign-in, and changePassword. The first button launches the sign-up modal and ajax is working well. The other two will not even launch, but the screen dims like it is about to. I have double and triple checked the html but cannot find the issue, and have not been able to find the issue.

**Comment by natGalynker** [comment]
```
<nav class="navbar navbar-default">
    <div class="container-fluid">
      <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#signUpModal">Sign UP</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#signInModal">Sign In</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#changePwModal">Change Password</button>
        <button type="button" class="btn btn-default navbar-btn" data-toggle="modal" data-target="#changePlayerModal">Change Player</button>
        <a class="navbar-brand" href="#">Tic Tac Toe</a>
      </div>
    </div>
  </nav>
  <div class="modal fade" id="signUpModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
          <h4 class="modal-title" id="myModalLabel">Sign Up</h4>
        </div>
        <div class="modal-body">
          <div class= "input-group">
            <form class="form-group" id="player-sign-up">
              <input type="text" class="form-control sign-in"
              name="credentials[player_name" placeholder="player name">
              <input type="text" class="form-control email"
              name="credentials[email]" placeholder="e-mail">
              <input type="password" class="form-control password"
              name="credentials[password]" placeholder="password">
              <input type="password" class="form-control confirm-pw"
              name="credentials[password_confirmation]"
              placeholder="confirm password">
              <div class="modal-footer">
                <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                <input type="submit" class="sign-button" name="submit" value="Sign up!">
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div class="modal fade" id="signInModal" tabindex="-1" role="dialog" aria-labelledby="thisModalLabel">
      <div class="modal-dialog" role="document">
        <div class="modal-content">
          <div class="modal-header">
            <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
            <h4 class="modal-title" id="thisModalLabel">Sign In</h4>
          </div>
          <div class="modal-body">
            <div class= "input-group">
              <form class="form-group" id="player-sign-in">
                <input type="text" class="form-control sign-in"
                name="credentials[player_name" placeholder="player name">
                <input type="text" class="form-control email"
                name="credentials[email]" placeholder="e-mail">
                <input type="password" class="form-control password"
                name="credentials[password]" placeholder="password">
                <div class="modal-footer">
                  <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                  <input type="submit" class="play-button" name="submit" value="Play Game!">
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="modal fade" id="changePwModal" tabindex="-1" role="dialog" aria-labelledby="changeModalLabel">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
          <h4 class="modal-title" id="changeModalLabel">Sign Up</h4>
        </div>
        <div class="modal-body">
          <div class= "input-group">
            <form class="form-group" id="player-change-pw">
              <input type="text" class="form-control sign-in"
              name="credentials[player_name" placeholder="player name">
              <input type="text" class="form-control email"
              name="credentials[email]" placeholder="e-mail">
              <input type="password" class="form-control password"
              name="credentials[password]" placeholder="password">
              <input type="password" class="form-control confirm-pw"
              name="credentials[password_confirmation]"
              placeholder="confirm password">
              <div class="modal-footer">
                <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                <input type="submit" class="change-password" name="submit" value="Update!">
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
```

**Comment by laurenfazah** [comment]
They had been launching before, hadn't they? If so, what has changed since then?

**Comment by natGalynker** [comment]
No I only was able to test sign-up. They all worked with Ajax before I changed the forms to be launched through the modal.

**Comment by RealWeeks** [comment]
@natGalynker so you getting a grey'ed out screen and a modal should appear but it's not, is that correct?  If it's what I think it is, it is a common bootstrap pitfall.

**Comment by natGalynker** [comment]
Yes, that is correct.

**Comment by RealWeeks** [comment]
This has to do with positing of parent elements.  Adjust the z-index, this is a case for using `!important;`

**Comment by natGalynker** [comment]
Thank you! I just looked up the docs for it. And will use that in CSS along with !important.

**Comment by RealWeeks** [comment]
You may want to be specific with your selector or you may alter some elements you didn't mean to.

**Comment by natGalynker** [comment]
Is by modal ID specific enough? Or is that too broad?

**Comment by natGalynker** [comment]
I both adjusted the z-index to 0 and also to -1. Neither worked. It made the modal which was working appear behind a grey transparent background. Another weird bug happened. Both the sign-up and the Change password modals are now working. But the others are still not launching

**Comment by natGalynker** [comment]
Because of the weird bug I tried copying and replacing the moral that's still unresponsive but that didn't work. It's the login one. So the rest of the Ajax calls depend on it for retrieving token/id

**Comment by laurenfazah** [comment]
I don't fully understand what you mean in your most recent comment. What state are your modals in now?

**Comment by natGalynker** [comment]
Issue fixed! I completely removed the unresponsive one along with the button that launched it then re implemented them and it fixed the issue. Patch isn't working now though, coming back with this error:
 XMLHttpRequest cannot load http://tic-tac-toe.wdibos.com/change-password1210. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8080' is therefore not allowed access. The response had HTTP status code 404. I can open a separate issue if that makes more sense

**Comment by laurenfazah** [comment]
Do you see the typo here:  http://tic-tac-toe.wdibos.com/change-password1210

**Comment by laurenfazah** [comment]
Also, this issue should be closed if it was resolved, and new issues should be opened separately.
